src/ctlib/ct.c

```diff
@@ -3942,6 +3965,52 @@ ct_cursor(CS_COMMAND * cmd, CS_INT type, CS_CHAR * name, CS_INT namelen, CS_CHAR
 		cursor->status.close      = TDS_CURSOR_STATE_UNACTIONED;
 		cursor->status.dealloc    = TDS_CURSOR_STATE_UNACTIONED;
 
+		if (option == CS_UNUSED	 ||  (option & CS_END) != 0) {
+			/* Try to figure out type of the cursor. */
+			char delimiter[] = "\n\t,.[]() ";
+			enum {
+				eBaseline,
+				eFor,
+				eForUpdate
+			} state = eBaseline;
+			char* savept = NULL;
+			char* s = text;
+
+			char* tok = get_next_tok(s, delimiter, &savept);
+			while (tok != NULL) {
+				s = savept;
+
+				if (strcasecmp(tok, "FOR") == 0) {
```

Comment:
`get_next_tok` is not terminating the tokens so if you have a query like `select * from table for update` `get_next_token` will return `for update`, not `for`.